File: src/main/python/basic/knapsack.py
#!/usr/bin/env python3.4


def knapsack(weight_arr, value_arr, w):
    if w == 0:
        return 0
    if w < 0:
        return -100000

    max_value = -1000
    for i in range(len(weight_arr)):
        rw = w - weight_arr[i]
        max_value = max(max_value, value_arr[i] + knapsack(weight_arr, value_arr, rw))
    return max_value


if __name__ == '__main__':
    val = [30, 100, 120]
    wt = [10, 20, 30]
    W = 50
    result_arr = list()
    result_arr.append(0)
    result_arr.append(-1)
    result_arr.append(-1)
    result_arr.append(-1)
    print (knapsack(wt, val, W))

File: src/main/perl/PerlProgramming/Knapsack.pl
#!/usr/bin/env perl -w

use diagnostics;
use warnings;
use strict;

sub Knapsack {

	my($result, $weight, $value, $w) = @_;
   
        my $max = 0;
	if(@$result[$w] >= 0) {
		return @$result[$w];
	}
	my $size = @$value;
	
	for(my $i = 0; $i < $size; $i++ ) {
		my $rw = $w - @$weight[$i];
		if($rw >= 0) {
			my $cur = &Knapsack(\@$result, \@$weight, \@$value, $rw) + @$value[$i];
			if($cur > $max) {
				$max = $cur;
			}
		}
	}
	@$result[$w] = $max;
	return $max;
}

my @result;
for(my $i = 0; $i <= 50; $i++) {
	$result[$i] = -1;
}
$result[0] = 0;

my @weight = (10, 20, 30);
my @value = (30, 100, 120);
my $start = time;
my $r = &Knapsack(\@result, \@weight, \@value, 50);
my $end = time - $start;
printf ("Result %s, Duration %d \n", $r,$end);
